File: public/sample_codes/algo/Fibonacci(DP)/fib.cpp
vector<int> dp;

int fib(int x, int ok=0){ 
    string nodeVal = "fib(x = " + to_string(x) + ")";
    tv.make(nodeVal, ok);

    if(x<2){
        tv.onHover("From base-case : " + to_string(x), ok);
        return x;
    } 

    if(dp[x] != -1){tv.onHover("From DP : " + to_string(dp[x]), ok); return dp[x];}

    int tmp = fib(x - 1, ok) + fib(x - 2, ok);
    dp[x] = tmp;
    tv.onHover(to_string(tmp), ok);
    return tmp;
}

//!----------------------------------------------------------------------------------

int main()
{
    //Call you rescursive function here @ top
    int n = 5;
    dp.resize(n+1, -1);
    fib(n);


    //Don't remove the below lines
    tv.buildTree();
    tv.printData();
}

File: public/sample_codes/algo/Fibonacci(DP)/fib.java
import java.util.Arrays;

public class Main {

    static TreeVisualizer tv = new TreeVisualizer();
    static int[] dp; 

    public static int fib(int x, int ok) {
        String nodeVal = "fib(x = " + x + ")";
        ok = tv.make(nodeVal, ok);

        if (x < 2) {
            tv.onHover("From base-case : " + x, ok);
            return x;
        }

        if (dp[x] != -1) {
            tv.onHover("From DP : " + dp[x], ok);
            return dp[x];
        }

        int tmp = fib(x - 1, ok) + fib(x - 2, ok);
        dp[x] = tmp; 
        tv.onHover(String.valueOf(tmp), ok);
        return tmp;
    }

    public static int fib(int x) {
        return fib(x, 0);
    }

    public static void main(String[] args) {
        int n = 5;
        
        dp = new int[n + 1];
        Arrays.fill(dp, -1);
        
        fib(n);

        tv.buildTree();
        tv.printData();
    }
}